### ISSUE TITLE
The _oldStyle decorator should let you set the bases of the created new-style class

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8308]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#8308|
|Type|enhancement|
|Created|2016-04-27 12:43:32Z|
|Branch|https://github.com/twisted/twisted/tree/oldstyle-bases-8308|

The current one only allows you to descend from object. When descending from the stdlib's old-style classes, hybrids will have to be made (```bases=(stdlibwhatever, object)```, and descending from stdlib's new-style classes will require descending from that class (```bases=(Exception,)``` for failure, for example).

<details><summary>Searchable metadata</summary>

```
trac-id__8308 8308
type__enhancement enhancement
reporter__hawkowl hawkowl
priority__normal normal
milestone__New_Style_Classes_Only New-Style Classes Only
branch__oldstyle_bases_8308 oldstyle-bases-8308
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1461761012673195 1461761012673195
changetime__1469694588435439 1469694588435439
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [47331]) Branching to oldstyle-bases-8308.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @hawkowl|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_4"></a>|@hawkowl commented|
|-|-|

Happy with this, please review.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="rodrigc's avatar" src="https://avatars.githubusercontent.com/u/1895943?s=50" width="50" height="50">](https://github.com/rodrigc)<a name="note_6"></a>|@rodrigc set owner to @hawkowl|
|-|-|

Test coverage of the diff is not 100%

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_7"></a>|@hawkowl removed owner|
|-|-|

100% now (with a new buildbot to cover the coverage).

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

We should have some machinery that hooks up the PR automatically...

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

Also, according to adi and rodrigc, we should be putting the ticket number first in the branch name, not last?

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @hawkowl|
|-|-|

Thanks hawkowl.  Clearly an oversight in its original functionality :).

  1. My primary question: why is this value not derived from `cls.__bases__`?  I'd rather just fix the implementation... tacking on `object` to `cls.__bases__` ought to eliminate the potential for errors and redundancy here.  It seems like the main reason to want that would be because `Failure` doesn't inherit from `Exception` normally but in the new-style future it needs to in order to be raiseable.  But in that specific case, no it doesn't!  :).  We already stipulate that the correct behavior on py3 is not `raise self`, but rather `self.raiseException()`.  The old behavior where `Failure` inherits from `Exception` for raiseability is a terrible bug that we should just get rid of.  It's a compat break, yes, but this env-var-to-change-behavior route we're following is specifically supposed to help with compat.
  2. `_oldStyle` doesn't give an example of its usage anywhere, and is documented as "a decorator" (which is subtly wrong since now it's a function that returns a decorator).  To alleviate this confusion - if we want to stick with this and have it take parameters, that is - it should probably have a first positional parameter that raises an explanatory exception if it's passed.  (What we really want of course is keyword-only parameters...)

I think the main thing here is to answer the question in my first feedback point and submit for re-review.  If you really feel like there's a good reason it should be this way, it probably ought to be in the docstring explaining why.  But, hopefully you can just re-do it with `__bases__` ;).

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_11"></a>|@hawkowl removed owner|
|-|-|

Okay, done [#11338](https://github.com/twisted/twisted/issues/11338), and I think that makes #2 unrequired.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

PR is here: https://github.com/twisted/twisted/pull/322

Three successful green checkmarks in a row. I can die happy.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_13"></a>|@glyph set owner to @hawkowl|
|-|-|

Awesome, thanks for responding to the feedback and simplifying.  I've made some comments in the PR, and it would be good if you could address them (since it would be even fewer lines of diff if you did) but all the churn I'm concerned about is private API, so you can always fix in a separate ticket if you'd like.  See inline github comments on the PR for more, but land at your pleasure!

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|GitHub set status to `closed`|
|-|-|

In changeset a43775dd9af5338f5f0076b2b87f8fe7327e8174
```
#!CommitTicketReference repository="" revision="a43775dd9af5338f5f0076b2b87f8fe7327e8174"
Merge pull request #322 from twisted/oldstyle-bases-8308

Author: hawkowl
Reviewer: glyph
Fixes: #8308
```
